zerver/migrations/0450_backfill_subscription_auditlogs.py

```diff
@@ -0,0 +1,151 @@
+from django.utils.timezone import now as timezone_now
+
+
+def backfill_missing_subscriptions(
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
+    """Backfill subscription realm audit log events for users which are
+    currently subscribed but don't have any, presumably due to some
+    historical bug.  This is important because those rows are
+    necessary when reactivating a user who is currently
+    soft-deactivated.
+
+    For each stream, we find the subscribed users who have no relevant
+    realm audit log entries, and create a backfill=True subscription
+    audit log entry which is the latest it could have been, based on
+    UserMessage rows.
+
+    """
+
+    Stream = apps.get_model("zerver", "Stream")
+    RealmAuditLog = apps.get_model("zerver", "RealmAuditLog")
+    Subscription = apps.get_model("zerver", "Subscription")
+    UserMessage = apps.get_model("zerver", "UserMessage")
+    Message = apps.get_model("zerver", "Message")
+
+    def get_last_message_id() -> int:
+        # We generally use this function to populate RealmAuditLog, and
+        # the max id here is actually system-wide, not per-realm.  I
+        # assume there's some advantage in not filtering by realm.
+        last_id = Message.objects.aggregate(Max("id"))["id__max"]
+        if last_id is None:
+            # During initial realm creation, there might be 0 messages in
+            # the database; in that case, the `aggregate` query returns
+            # None.  Since we want an int for "beginning of time", use -1.
+            last_id = -1
+        return last_id
+
+    for stream in Stream.objects.all():
+        with transaction.atomic():
+            subscribed_user_ids = set(
+                Subscription.objects.filter(recipient_id=stream.recipient_id).values_list(
+                    "user_profile_id", flat=True
+                )
+            )
+            user_ids_in_audit_log = set(
+                RealmAuditLog.objects.filter(
+                    realm=stream.realm,
+                    event_type__in=[
+                        301,  # RealmAuditLog.SUBSCRIPTION_CREATED
+                        302,  # RealmAuditLog.SUBSCRIPTION_ACTIVATED
+                        303,  # RealmAuditLog.SUBSCRIPTION_DEACTIVATED
+                    ],
+                    modified_stream=stream,
+                )
+                .distinct("modified_user_id")
+                .values_list("modified_user_id", flat=True)
+            )
+
+            user_ids_missing_events = subscribed_user_ids - user_ids_in_audit_log
+            if not user_ids_missing_events:
+                continue
+
+            last_message_id = get_last_message_id()
+            now = timezone_now()
+            backfills = []
+            for user_id in sorted(user_ids_missing_events):
+                print(
+                    f"Backfilling subscription event for {user_id} in {stream.name} in {stream.realm.string_id}"
```

Comment:
I think we want to log the stream ID instead of the name, both for clarity and also for privacy reasons.